File: algorithms/[1232]缀点成线/solution.py
# topics = ["数学"]

from typing import List


class Solution:
    def checkStraightLine(self, coordinates: List[List[int]]) -> bool:
        """
        判断向量共线
        time O(n), space O(1), n 为数组长度
        """
        [x0, y0] = coordinates[0]
        [x1, y1] = coordinates[1]

        for [x, y] in coordinates:
            if (x1 - x0) * (y - y0) - (y1 - y0) * (x - x0):
                return False

        return True

File: algorithms/[1232]缀点成线/solution.ts
// topics = ["数学"]

/**
 * 判断斜率
 * time O(n), space O(1), n 为数组长度
 */
function checkStraightLine(coordinates: number[][]): boolean {
  const [x0, y0] = coordinates[0];
  const [x1, y1] = coordinates[1];

  // 判断是否斜率为 0
  if (coordinates.every(([x, _]) => x === x1)) {
    return true;
  }

  const a = (y1 - y0) / (x1 - x0);
  const b = y0 - a * x0;

  return coordinates.every(([x, y]) => y === a * x + b);
}
